pkg/controller/identity/iamrole/controller.go

```diff
@@ -107,38 +109,39 @@ func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.E
 	}
 
 	if err != nil {
-		return managed.ExternalObservation{}, errors.Wrapf(err, errGet, cr.Spec.RoleName)
+		return managed.ExternalObservation{}, errors.Wrapf(err, errGet, meta.GetExternalName(cr))
 	}
 
 	cr.SetConditions(runtimev1alpha1.Available())
 
-	cr.UpdateExternalStatus(*observed.Role)
+	iam.UpdateRoleExternalStatus(cr, *observed.Role)
 
 	return managed.ExternalObservation{
 		ResourceExists: true,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRole)
+	cr, ok := mgd.(*v1beta1.IAMRole)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
 	cr.Status.SetConditions(runtimev1alpha1.Creating())
 
-	req := e.client.CreateRoleRequest(&awsiam.CreateRoleInput{
-		RoleName:                 aws.String(cr.Spec.RoleName),
-		AssumeRolePolicyDocument: aws.String(cr.Spec.AssumeRolePolicyDocument),
-		Description:              aws.String(cr.Spec.Description),
-	})
+	name := meta.GetExternalName(cr)
+	if name == "" {
+		name = cr.GetGenerateName()
+	}
+
+	req := e.client.CreateRoleRequest(iam.GenerateCreateRoleInput(name, &cr.Spec.ForProvider))
 
 	result, err := req.Send(ctx)
 	if err != nil {
 		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
 	}
 
-	cr.UpdateExternalStatus(*result.Role)
+	iam.UpdateRoleExternalStatus(cr, *result.Role)
```

Comment:
We typically would update the status in the subsequent `Observe`.

Ref: https://github.com/crossplane/stack-aws/blob/94baed407fc79f3a083c409b42542a00f7f0eb74/pkg/controller/database/rdsinstance.go#L140